pkg/agent/client/endpoint_resolver.go

```diff
@@ -0,0 +1,218 @@
+	// serviceLister is used to retrieve the antrea Service when selecting one of the Service's Endpoints.
+	serviceLister       corev1listers.ServiceLister
+	serviceListerSynced cache.InformerSynced
+	// endpointLister is used to select and Endpoint for the antrea Service. This Endpoint will
+	// be used to access the Service without depending on the availability of a proxy (whether
+	// kube-proxy or AntreaProxy).
+	endpointsLister       corev1listers.EndpointsLister
+	endpointsListerSynced cache.InformerSynced
+	queue                 workqueue.RateLimitingInterface
+	listeners             []Listener
+	endpointURL           atomic.Pointer[url.URL]
+}
+
+func NewEndpointResolver(
+	namespace, serviceName string,
+	servicePort int32,
+	serviceInformer coreinformers.ServiceInformer,
+	endpointsInformer coreinformers.EndpointsInformer,
+) *EndpointResolver {
+	key := namespace + "/" + serviceName
+	controllerName := fmt.Sprintf("ServiceEndpointResolver-%s", key)
+	resolver := &EndpointResolver{
+		name:                  controllerName,
+		namespace:             namespace,
+		serviceName:           serviceName,
+		servicePort:           servicePort,
+		serviceLister:         serviceInformer.Lister(),
+		serviceListerSynced:   serviceInformer.Informer().HasSynced,
+		endpointsLister:       endpointsInformer.Lister(),
+		endpointsListerSynced: endpointsInformer.Informer().HasSynced,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(100*time.Millisecond, 5*time.Second), controllerName),
+	}
+	serviceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			if service, ok := obj.(*corev1.Service); ok {
+				return service.Namespace == namespace && service.Name == serviceName
+			}
+			if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				if service, ok := tombstone.Obj.(*corev1.Service); ok {
+					return service.Namespace == namespace && service.Name == serviceName
+				}
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				resolver.queue.Add(key)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				// This should not happen: both objects should be Services in the
+				// update event handler.
+				oldSvc, ok := oldObj.(*corev1.Service)
+				if !ok {
+					return
+				}
+				newSvc, ok := newObj.(*corev1.Service)
+				if !ok {
+					return
+				}
+				// We do not care about potential Status updates.
+				if reflect.DeepEqual(newSvc.Spec, oldSvc.Spec) {
+					return
+				}
+				resolver.queue.Add(key)
+			},
+			DeleteFunc: func(obj interface{}) {
+				resolver.queue.Add(key)
+			},
+		},
+	})
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			// The Endpoints resource for a Service has the same name as the Service.
+			if service, ok := obj.(*corev1.Service); ok {
```

Comment:
should it be Endpoints?

Reply:
Thanks 😊  that explains why the tests are failing...